### PR TITLE
Gradle optimizations

### PR DESCRIPTION
**Pull request by tylerbertrand**
## Description

Ported recent build improvements from Apache Solr ([#1319](https://github.com/apache/solr/pull/1319) and [#1345](https://github.com/apache/solr/pull/1345)) with some modifications.

In addition, the following build issues were addressed:
1. Task `validateSourcePatterns` was including the `.gradle` directory for both `buildSrc` and `dev-tools/missing-doclet` as inputs, causing this task to not be cacheable.
   * Resolved by excluding all `.gradle` directories as inputs to `validateSourcePatterns`, not just the `.gradle` directory for the root project.
2. With the changes to `copyTestResources`/`processTestResources`, the test `TestCustomAnalyzer.testStopWordsFromFile` was failing because `teststop.txt` was no longer at the root of the classes directory.
   * Resolved by following the convention set by other tests: providing the path to the file with `this.getDataPath("teststop.txt").toString()`

## Improvements

Improved overall build time:
  * [Build scan timeline summary before changes](https://scans.gradle.com/s/wyryx5onlnlls#timeline)
  * [Build scan timeline summary after changes](https://scans.gradle.com/s/qupifnuiysdfq#timeline)

Reduced number of executed cacheable tasks (cache misses):
  * [Build scan before changes](https://scans.gradle.com/s/wyryx5onlnlls/timeline?cacheability=cacheable&outcome=success,failed&sort=longest)
  * [Build scan after changes](https://scans.gradle.com/s/qupifnuiysdfq/timeline?cacheability=cacheable&outcome=success,failed)

Improved the number of cacheable tasks:
  * [Non-cacheable tasks before changes](https://scans.gradle.com/s/wyryx5onlnlls/timeline?cacheability=any-non-cacheable&outcome=success,failed&sort=longest)
  * [Non-cacheable tasks after changes](https://scans.gradle.com/s/qupifnuiysdfq/timeline?cacheability=any-non-cacheable&outcome=success,failed)

Reduced the number of eagerly created tasks:
  * [Eagerly created tasks before changes](https://scans.gradle.com/s/wyryx5onlnlls/performance/configuration#summary-task-created-immediately)
  * [Eagerly created tasks after changes](https://scans.gradle.com/s/qupifnuiysdfq/performance/configuration#summary-task-created-immediately)

### CODE REVIEW

File: gradle/validation/jar-checks.gradle [hunk 1]
```diff
@@ -231,7 +238,8 @@ subprojects {
           }
         }
       }
-
+      def f = new File(project.buildDir.path + "/" + outputFileName)
+      f.text = errors
```
Comment (dweiss) [inline gradle/validation/jar-checks.gradle:242]: This uses local encoding. Should use UTF8.

**Comment by dweiss** [comment]
I don't like how gradle requires that extra verbosity (command line providers) for the sake of achieving the optimal task-caching behavior, but I'm fine with these changes and I think I understand what's going on.

File: gradle/validation/jar-checks.gradle (hunk 1 above)
Comment (uschindler) [reply gradle/validation/jar-checks.gradle:242]: This was also part of the Solr PR and was overlooked there.

In general I don't understand why this was added at all. What is the output file used for at all?

**Review by uschindler — COMMENTED** [review]
I agree with the changes in general, but following Dawid's UTF-8 forbiddenapis detection I'd like to first know why the additional output file is needed at all.

**Comment by dweiss** [comment]
An output file is required for caching - otherwise there is no link between inputs-and-outputs (up-to-date checks). A task without outputs is never skipped in incremental builds (it's never 'up to date').

File: gradle/validation/jar-checks.gradle (hunk 1 above)
Comment (tylerbertrand) [reply gradle/validation/jar-checks.gradle:242]: Dawid is correct, for tasks to take advantage of incremental building and the build cache, they're required to have an output.

Comment (tylerbertrand) [reply gradle/validation/jar-checks.gradle]: Fixed

**Review by uschindler — APPROVED** [review]
Looks fine.

**Comment by clayburn** [comment]
Hi, is there anything blocking merge here?

**Comment by dweiss** [comment]
No, not really. I've been out of reach recently - sorry. I'll merge it in.

**Comment by dweiss** [comment]
I'll apply this to 9x as well.

**Comment by risdenk** [comment]
Closes https://github.com/apache/lucene/issues/12145

**Comment by uschindler** [comment]
Unfortunately the changes applied here for the file `jar-checks.gradle` (the new build outputs) cause random build warnings, see https://github.com/apache/lucene/issues/13695.

The problem with all the tasks in `jar-checks.gradle` are that they share internal state between the tasks outside of files. When one of the tasks do not run, some info is missing causing random license warnings.

As quick workaround I will remove the output of the task `validateJarLicenses` in file `jar-checks.gradle`.

**Pull request merged**